File: client/src/app/core/common/paged-result.ts
export interface PagedResult<T> {
  items: T[];
  totalCount: number;
  page: number;
  pageSize: number;
  /** Server-computed; falls back to 0 when totalCount or pageSize is 0. */
  totalPages?: number;
}

/** Common page-size choices for our list views. */
export const PAGE_SIZE_OPTIONS = [10, 20, 50, 100] as const;
export type PageSize = (typeof PAGE_SIZE_OPTIONS)[number];
export const DEFAULT_PAGE_SIZE: PageSize = 20;

File: src/DailyTracker.Application/Dtos/Common/PagedResult.cs
namespace DailyTracker.Application.Dtos.Common;

/// <summary>
/// Standard envelope for paginated list endpoints.
/// </summary>
public sealed record PagedResult<T>(
    IReadOnlyList<T> Items,
    int TotalCount,
    int Page,
    int PageSize)
{
    public int TotalPages => PageSize > 0 ? (int)Math.Ceiling((double)TotalCount / PageSize) : 0;
}

/// <summary>
/// Common query parameters for paginated endpoints. Defaults and clamping live
/// here so every controller doesn't reinvent the same guard rails.
/// </summary>
public sealed record PageQuery
{
    private const int DefaultPageSize = 500;
    private const int MaxPageSize = 1000;

    public int Page { get; init; } = 1;
    public int PageSize { get; init; } = DefaultPageSize;

    public int NormalizedPage => Page < 1 ? 1 : Page;
    public int NormalizedPageSize => PageSize switch
    {
        < 1 => DefaultPageSize,
        > MaxPageSize => MaxPageSize,
        _ => PageSize,
    };
    public int Skip => (NormalizedPage - 1) * NormalizedPageSize;
}
